Modules/Loadable/VolumeRendering/MRMLDM/vtkMRMLVolumeRenderingDisplayableManager.cxx

```diff
@@ -503,67 +552,86 @@ void vtkMRMLVolumeRenderingDisplayableManager::vtkInternal::AddDisplayNode(vtkMR
 //---------------------------------------------------------------------------
 void vtkMRMLVolumeRenderingDisplayableManager::vtkInternal::RemoveDisplayNode(vtkMRMLVolumeRenderingDisplayNode* displayNode)
 {
-  PipelinesCacheType::iterator pipelineIt = this->DisplayPipelines.find(displayNode);
+  if (!displayNode)
+    {
+    return;
+    }
+  
+  PipelineListType::iterator pipelineIt = this->GetPipelineIt(displayNode);
   if (pipelineIt == this->DisplayPipelines.end())
     {
     return;
     }
+  this->RemovePipelineIt(pipelineIt);
+}
 
-  const Pipeline* pipeline = pipelineIt->second;
-
-  if ( displayNode->IsA("vtkMRMLCPURayCastVolumeRenderingDisplayNode")
-    || displayNode->IsA("vtkMRMLGPURayCastVolumeRenderingDisplayNode") )
+//---------------------------------------------------------------------------
+vtkMRMLVolumeRenderingDisplayableManager::vtkInternal::PipelineListType::iterator
+vtkMRMLVolumeRenderingDisplayableManager::vtkInternal::RemovePipelineIt(PipelineListType::iterator pipelineIt)
+{
+  if (pipelineIt == this->DisplayPipelines.end())
     {
-    // Remove volume actor from renderer and local cache
-    this->External->GetRenderer()->RemoveVolume(pipeline->VolumeActor);
+    // already removed
+    return this->DisplayPipelines.end();
     }
+
+  Pipeline* pipeline = *pipelineIt;
+
+  if (pipeline)
+  {
+    if (pipeline->VolumeActor)
+      {
+      this->External->GetRenderer()->RemoveVolume(pipeline->VolumeActor);
+      }
+
 #if VTK_MAJOR_VERSION >= 9 || (VTK_MAJOR_VERSION >= 8 && VTK_MINOR_VERSION >= 2)
-  else if (displayNode->IsA("vtkMRMLMultiVolumeRenderingDisplayNode"))
-    {
-    const PipelineMultiVolume* pipelineMulti = dynamic_cast<const PipelineMultiVolume*>(pipeline);
+    PipelineMultiVolume* pipelineMulti = dynamic_cast<PipelineMultiVolume*>(pipeline);
     if (pipelineMulti)
       {
       // Remove volume actor from multi-volume actor collection
       this->MultiVolumeMapper->RemoveInputConnection(pipelineMulti->ActorPortIndex, 0);
       this->MultiVolumeActor->RemoveVolume(pipelineMulti->ActorPortIndex);
 
+      // Decrease next actor port index
```

Comment:
@cpinter Should not we update port numbers in other pipelines (decrement the port index of those pipelines which come after the deleted actor)?

Reply:
Yes, based on the code I think you are right.

That would mean that if I show two volumes, and then I hide the first one, then unexpected things should happen. I tested it with four volumes, showing/hiding them in lots of different ways then adjusting settings, and it seems to work. Maybe it's just a coincidence. However, what you suggest seems valid.